src/components/graph/ExerciseGraphLayout.ts

```diff
@@ -0,0 +1,337 @@
+            )
+            return true
+          }
+        }
+      }
+    }
+    return false
+  }
+
+  /**
+   * computeNodeDepths
+   * This computes the depth of all nodes, first exercises then prerequisite concepts
+   * that are not yet represented in an exercise.
+   */
+  private computeNodeDepths(): void {
+    const inDegrees = this.findIndegreeOfExercises()
+    const sourceExercises = this.findSourceExercises(inDegrees)
+    const exerciseDepths = this.findExerciseDepths(sourceExercises)
+    const maximumDepth = Math.max.apply(null, exerciseDepths)
+    this.exerciseLayers = this.constructExerciseLayers(
+      exerciseDepths,
+      maximumDepth
+    )
+    this.conceptLayers = this.constructConceptLayers(this.exerciseLayers)
+  }
+
+  /**
+   * Use the adjacency matrix to find the in-degree of each
+   * node (the number of edges pointing to the node)
+   *
+   * Return an array of numbers, where each index (representing
+   * an ExerciseGraphNode) holds the indegree value
+   */
+  private findIndegreeOfExercises(): Array<number> {
+    const indegrees = new Array(this.adjacency.length).fill(0)
+    for (const tos of this.adjacency) {
+      for (const to of tos) {
+        indegrees[to] += 1
+      }
+    }
+    return indegrees
+  }
+
+  /**
+   * Nodes with inDegree of 0 are source nodes (roots) of the graph
+   */
+  private findSourceExercises(inDegrees: Array<number>): Array<number> {
+    const sources: Array<number> = []
+    inDegrees.forEach((inDegree, i) => {
+      if (inDegree === 0) sources.push(i)
+    })
+    return sources
+  }
+
+  /**
+   * Perform a recursive search to find the length of longest path
+   * to each exercise from any source.  This places an exercise
+   * after all of its prerequisites.
+   */
+  private findExerciseDepths(sources: Array<number>): Array<number> {
+    const maxDepths = new Array(this.adjacency.length).fill(0)
+    for (const source of sources) {
+      const traverse = (index: number, depth = 0): void => {
+        if (depth > maxDepths[index]) {
+          maxDepths[index] = depth
+        }
+
+        const children = this.adjacency[index]
+        children.forEach((child) => traverse(child, depth + 1))
+      }
+      traverse(source)
+    }
+    return maxDepths
+  }
+
+  /**
+   * order the exercises by depth then index into layers
+   */
+  private constructExerciseLayers(
+    exerciseDepths: Array<number>,
+    maxDepth: number
+  ): Array<Array<ExerciseGraphNode>> {
+    const exerciseLayers = new Array(maxDepth + 1)
+
+    exerciseDepths.forEach((nodeDepth, nodeIndex) => {
+      exerciseLayers[nodeDepth] = exerciseLayers[nodeDepth] ?? []
+      exerciseLayers[nodeDepth].push(this.graph.lookupByIndex.get(nodeIndex))
+    })
+
+    return exerciseLayers
+  }
+
+  /**
+   * order the concepts without exercises by the depth of their dependent exercise
+   */
+  private constructConceptLayers(
+    exerciseLayers: Array<Array<ExerciseGraphNode>>
```

Comment:
```suggestion
    exerciseLayers: ExerciseGraphNode[][]
```

You could also extract this as a type:

```typescript
type ExerciseLayer = ExerciseGraphNode[]
type ExerciseLayers = ExerciseLayer[]